packages/react-search-ui/src/withSearch.js

```diff
@@ -1,23 +1,81 @@
 import React from "react";
 
-import SearchConsumer from "./SearchConsumer";
+import SearchContext from "./SearchContext";
+
+function buildContextForProps(context) {
+  return {
+    ...context.driver.getState(),
+    ...context.driver.getActions()
+  };
+}
+
+/* For a given object, pluck out the key/value pairs matching the keys
+provided in the uses parameter */
+function giveMeJustWhatINeeded(stateOrContext, uses) {
+  if (uses.length === 0) return stateOrContext;
+  return uses.reduce((acc, use) => {
+    if (!stateOrContext.hasOwnProperty(use)) return acc;
+    return {
+      ...acc,
+      [use]: stateOrContext[use]
+    };
+  }, {});
+}
 
 /**
- * This is a Higher Order Component that is used to expose (as `props`) all
- * state and Actions provided by the SearchProvider to "container"
- * components.
+ * This is a Higher Order Component that wraps a component and injects state and actions from Search UI, effectively
+ * "connecting" it to Search UI.
  *
  * `mapContextToProps` can be used to manipulate actions and state from context
  * before they are passed on to the container.
+ *
+ * Components using `withSearch` will be "Pure" components. A "shouldComponentUpdate" is implemented
+ * below to ensure that components only render when state has changed.
+ *
+ * It is important to understand the implications of using a PureComponent, as described here:
+ * https://reactjs.org/docs/optimizing-performance.html#examples
+ *
+ * @param Array[String] uses An array of state or action values to be injected as props into the component. Provide
+ * an empty array for all state and actions. This is not desirable because it can have bad performance characteristics
+ * as the component would then render every time state is updated in Search UI.
+ * @param Function Component
  */
-export default function withSearch(Component) {
-  return function WithSearch(props) {
-    const { mapContextToProps = context => context, ...rest } = props;
-
-    return (
-      <SearchConsumer>
-        {context => <Component {...mapContextToProps(context)} {...rest} />}
-      </SearchConsumer>
-    );
-  };
+export default function withSearch(uses = [], Component) {
+  class WithSearch extends React.PureComponent {
+    constructor() {
+      super();
+      this.state = {};
+      this.uses = uses;
+    }
+
+    componentWillMount() {
```

Comment:
componentWillMount is deprecated and will be removed - https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html

Reply:
Good catch.